src/lib.rs

```diff
@@ -4595,3 +4596,292 @@ mod tests {
+        //     let insufficient_bytes = bytes_len < base_layout.size();
+        //     // 2. performing the cast would misalign type:
+        //     let misaligned = (match cast_type {
+        //         _CastType::_Prefix => 0,
+        //         _CastType::_Suffix => bytes_len,
+        //     } + addr)
+        //         % base_layout.align()
+        //         != 0;
+
+        //     assert!(insufficient_bytes || misaligned);
+        // }
+    }
+
+    // Validates that `validate_cast_and_convert_metadata` satisfies its own
+    // documented safety postconditions, and also a few other properties
+    // that aren't documented but we want to guarantee anyway.
+    fn validate_behavior(
+        (layout, addr, bytes_len, cast_type): (DstLayout, usize, usize, _CastType),
+    ) {
+        if let Some((elems, split_at)) =
+            layout._validate_cast_and_convert_metadata(addr, bytes_len, cast_type)
+        {
+            let (size_info, align) = (layout._size_info, layout._align);
+            // If this is a sized type (no trailing slice), then `elems` is
+            // meaningless, but in practice we set it to 0. Callers are not
+            // allowed to rely on this, but a lot of math is nicer if
+            // they're able to, and some callers might accidentally do that.
+            let sized = matches!(layout._size_info, SizeInfo::Sized { .. });
+            assert!(!(sized && elems != 0), "{}", "");
+
+            let resulting_size = match layout._size_info {
+                SizeInfo::Sized { _size } => _size,
+                SizeInfo::SliceDst(TrailingSliceLayout {
+                    _offset: offset,
+                    _elem_size: elem_size,
+                }) => {
+                    let padded_size = |elems| {
+                        let max_total_bytes = bytes_len - (bytes_len % align);
+                        let max_slice_and_padding_bytes = max_total_bytes - offset;
+                        let slice_bytes =
+                            max_slice_and_padding_bytes - (max_slice_and_padding_bytes % elem_size);
+
+                        // Proves that the above is equivalent to `slice_bytes =
+                        // elems * elem_size`.
+                        assert_eq!(slice_bytes % elem_size, 0);
+                        assert_eq!(slice_bytes / elem_size, elems);
+
+                        let without_padding = offset + slice_bytes;
+
+                        // let without_padding = offset + elems * elem_size;
+                        without_padding
+                            + util::core_layout::_padding_needed_for(without_padding, align)
+                    };
+
+                    let resulting_size = padded_size(elems);
+                    // Test that `validate_cast_and_convert_metadata`
+                    // computed the largest possible value that fits in the
+                    // given range.
+                    // assert!(padded_size(elems + 1) > bytes_len, "{}", "");
```

Comment:
Kani finds a counter-example for this assertion, where `padded_size(elems + 1)` *equals* `bytes_len`.